File: leetcode/0020-valid-parentheses/isValid.go
package leetcode

func isValid(s string) bool {
	var stack = []rune{}
	var parenthesesMap = map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	for _, parentheses := range s {
		matchParenthese, ok := parenthesesMap[parentheses]
		if !ok {
			stack = append(stack, parentheses)
			continue
		}

		if len(stack) < 1 || stack[len(stack)-1] != matchParenthese {
			stack = append(stack, parentheses)
			continue
		}

		stack = stack[:len(stack)-1]
	}

	return len(stack) == 0
}

File: leetcode/0020-valid-parentheses/isValid.ts
function isValid(s: string): boolean {
  const stack = [];
  const parenthesesMap = new Map([
    [")", "("],
    ["}", "{"],
    ["]", "["],
  ]);

  for (const parentheses of s) {
    var matchParentheses = parenthesesMap.get(parentheses);

    if (!matchParentheses) {
      stack.push(parentheses);
      continue;
    }

    var lastItem = stack[stack.length - 1];
    if (lastItem !== matchParentheses) {
      stack.push(parentheses);
      continue;
    }

    stack.pop();
  }

  return stack.length == 0;
}

export default isValid;
